core/drivers/stm32mp15_huk.c

```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2022, Linaro Limited
+ * Copyright (c) 2022, Foundries.io Limited
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <crypto/crypto.h>
+#include <drivers/stm32_bsec.h>
+#include <kernel/tee_common_otp.h>
+#include <mempool.h>
+#include <platform_config.h>
+#include <string.h>
+#include <string_ext.h>
+
+static bool stm32mp15_huk_init;
+
+static TEE_Result stm32mp15_read_uid(uint32_t *uid)
+{
+	uint32_t *q = uid;
+
+	if (stm32_bsec_shadow_read_otp(q++, UID0_OTP))
```

Comment:
For a real locked state result in `tee_otp_get_hw_unique_key()`, you would need here to call `stm32mp15_read_otp()` for each `OTP_UIDx`.
That's not a mandatory fix. Locked/unlocked state only triggers a console trace.

Reply:
um, also on the UIDs? Is there a risk to this values changing between reads?

Reply:
Reading again the spec, there are some useful details.

IIUC the UID words (words 13-15: ID0, ID1, ID2) are always permantent locked but the shadow memory for UID words can be sticky locked of not. To ensure you use fuses definitive values, optee should test sw_lock is set, or more simply lock sw_lock for thoses words.

Regdarding where HUK will be placed (so-called upper OTPs), when the device is in "secure-closed" state, the shadow memory for these words is fully locked (no reload/reprogam) so no need to shadow and to test locks. When not "secure-closed" state, the several word locks may be released. In this case, using definitive fuses requires the sequence you implmented (shadow value + test locks).

Reply:
thanks a lot for checking these details @etienne-lms . I have incorporated your additions.